rfcs/20190729-keras-preprocessing-redesign.md

```diff
@@ -0,0 +1,351 @@
+
+- It supports GPU prefetching
+- It supports distribution via the Distribution Strategies API
+
+Meanwhile, `keras.preprocessing` is a major API for data loading and preprocessing in Keras. It is based
+on Numpy and Scipy, and it produces instances of the `keras.utils.Sequence` class, which are finite-length,
+resettable Python generators that yield batches of data.
+
+Some features of `keras.preprocessing` are highly useful and don't have straightforward equivalents in `tf.data`
+(in particular image data augmentation and dynamic time series iteration).
+
+Ideally, the utilities in `keras.preprocessing` should be made compatible with `tf.data`.
+This presents the opportunity to improve on the existing API. In particular we don't have good support
+for image segmentation use cases today.
+
+Some features are also being supplanted by [preprocessing layers](https://github.com/keras-team/governance/blob/master/rfcs/20190502-preprocessing-layers.md), in particular text processing. 
+As a result we may want move the current API to an API similar to Layers.
+
+
+## Goals
+
+- Unify "keras.preprocessing" and the recently-introduced [Preprocessing Layers API](https://github.com/keras-team/governance/blob/master/rfcs/20190502-preprocessing-layers.md).
+- Make all features of `keras.preprocessing` compatible with `tf.data`.
+- As a by-product, add required ops to TensorFlow (`tf.image`).
+
+
+## Proposed changes at a high-level
+
+
+- Deprecate `ImagePipelineGenerator` in favor of new `ImagePipeline` class similar to a `Sequential` model.
+- Inherits from `keras.layers.PreprocessingLayer` for all image transformations.
+- Deprecate `Tokenizer` class in favor of `TextVectorization` preprocessing layer.
+- Replace `TimeseriesGenerator` with a function-based API.
+
+
+## Detailed API changes
+
+
+### ImagePipeline
+
+#### Constructor
+
+`ImagePipeline` inherits from `keras.model.Sequential` and takes a list of layers as inputs. In the future it will inherit from `PreprocessingStage`.
+
+`ImagePipeline` is a preprocessing layer that encapsulate a series of image transformations. Since some of these transformations may be trained (featurewise normalization), it exposes the method `adapt`, like all other preprocessing layers.
+
+
+```python
+
+class ImagePipeline(Sequential):
+
+    def __init__(self, layers:List[Layer]):
+        ...
+```
+
+#### Example usage
+
+```python
+preprocessor = ImagePipeline([
+    RandomFlip(horizontal=True),
+    RandomRotation(0.2, fill_mode='constant'),
+    RandomZoom(0.2, fill_mode='constant'),
+    RandomTranslation(0.2, fill_mode='constant'),
+    Normalization(),  # This is the same Normalization introduced in preprocessing layers
+])
+preprocessor.adapt(sample_data)  # optional step in case the object needs to be trained
+
+dataset = preprocessor.from_directory(dir_name, image_size=(512, 512))
+model.fit(dataset, epochs=10)
+```
+
+#### Methods
+
+```python
+def from_directory(
```

Comment:
To clarify-- these are now methods of an instantiated ImagePipeline, not standalones?

Reply:
Yes, these are instance methods, not standalone functions.